File: solutions/python/rna-transcription/1/rna_transcription.py
def to_rna(dna_strand):
    """
    Transcribe a DNA strand into its corresponding RNA strand using RNA Interference principles.

    The transcription is performed by replacing each nucleotide in the DNA strand with
    its RNA complement:

    DNA -> RNA
    G   -> C
    C   -> G
    T   -> A
    A   -> U

    Parameters:
    dna_strand (str): A string representing the DNA strand, composed of 'A', 'C', 'G', and 'T'.

    Returns:
    str: A string representing the corresponding RNA strand.
    
    Raises:
    ValueError: If the DNA strand contains invalid characters.
    """
    transcription_map = {
        'G': 'C',
        'C': 'G',
        'T': 'A',
        'A': 'U'
    }

    try:
        # Translate each nucleotide using the transcription map
        return ''.join(transcription_map[nucleotide] for nucleotide in dna_strand)
    except KeyError as e:
        raise ValueError(f"Invalid nucleotide found in DNA strand: {e.args[0]}")

File: solutions/typescript/rna-transcription/1/rna-transcription.ts
const complementMap: Record<'G' | 'C' | 'T' | 'A', 'C' | 'G' | 'A' | 'U'> = {
  G: 'C',
  C: 'G',
  T: 'A',
  A: 'U'
};

/**
 * Transcribes a DNA strand into its RNA complement using RNA Interference mapping.
 * @param {string} dna - The DNA sequence to transcribe.
 * @returns {string} - The corresponding RNA sequence.
 * @throws {Error} - If the DNA sequence contains invalid nucleotides.
 */
export function toRna(dna: string): string {
  return dna
    .split('')
    .map((nucleotide: string) => {
      const rna = complementMap[nucleotide as keyof typeof complementMap];
      if (rna === undefined) {
        // Unified error message for any invalid input
        throw new Error('Invalid input DNA.');
      }
      return rna;
    })
    .join('');
}